lib/utf8_reader.rb

```diff
@@ -0,0 +1,66 @@
+##
+# This class Reads individual characters from files using UTF-8 encoding
+# This deals with two main concerns:
+#   - Variable byte length of characters
+#   - Reducing the number of read operations by loading bytes in chunks
+
+class FormatParser::UTF8Reader
```

Comment:
you mean the `safe_read` method for dealing with errors raised in EOF or the `IOConstraint` class?

as for `IOConstraint`, it is in use by default, as that's what a parser hands over to this validator.
`safe_read` can be used here if you think it's necessary, as long as we rescue those to new possible errors down the line.

Any thoughts?

Reply:
The former. If we already have the logic to read N bytes from various sources and deal with potential errors and such, I was curious why also (re)implementing this here. Probably I'm missing the larger context of how `safe_read` works 🤔

Reply:
the point of this UTF8 reader is: we don't actually know how many bytes we want to read, since chars can have different lenghts - from 1 to 4. This reader abstracts that away and gives us back chars one by one, no matter their size.

internally it could use safe_read, but that would only deal with `EOF`, which is not a problem at all here. 

```
def safe_read(io, n)
    raise ArgumentError, 'Unbounded reads are not supported' if n.nil?
    buf = io.read(n)

    raise InvalidRead, "We wanted to read #{n} bytes from the IO, but the IO is at EOF" unless buf
    raise InvalidRead, "We wanted to read #{n} bytes from the IO, but we got #{buf.bytesize} instead" if buf.bytesize != n

    buf
  end

```

Reply:
Right, that makes more sense 👍